README.md
# paytm

A Flutter plugin to use the Paytm as a payment gateway for accepting online payments in Flutter app.

## For Testing/Staging credentials make sure appInvokeEnabled set to FALSE or Paytm APP is not installed on the device.


## Example App in iOS
<!-- ![IMG_0CA99F9C709C-1](https://user-images.githubusercontent.com/25786428/82787888-07fbc180-9e85-11ea-87cb-754c6155b1d3.jpeg) -->
<img src="https://user-images.githubusercontent.com/25786428/82787888-07fbc180-9e85-11ea-87cb-754c6155b1d3.jpeg" width="300">


## First of all get Credentials from Paytm
Plugin will only work with Production Keys
[https://dashboard.paytm.com/next/apikeys](https://dashboard.paytm.com/next/apikeys)


## Let’s begin

iOS Configuration:

In case merchant don’t have callback URL, Add an entry into Info.plist

1. LSApplicationQueriesSchemes(Array) Item 0 (String)-> paytm
![iosInvoke](https://user-images.githubusercontent.com/25786428/82787548-45138400-9e84-11ea-835f-caa0701728cb.png)

2. Add a URL Scheme “paytm”+”MID”
![app-invoke-ios-inti](https://user-images.githubusercontent.com/25786428/82787531-3c22b280-9e84-11ea-9923-c18f2bc904de.png)
 

## Start Payment
```
  void generateTxnToken(int mode) async {
    
        String orderId = DateTime.now().millisecondsSinceEpoch.toString();

        String callBackUrl = (testing
                ? 'https://securegw-stage.paytm.in'
                : 'https://securegw.paytm.in') +
            '/theia/paytmCallback?ORDER_ID=' +
            orderId;

        //Host the Server Side Code on your Server and use your URL here. The following URL may or may not work. Because hosted on free server.
        //Server Side code url: https://github.com/mrdishant/Paytm-Plugin-Server
        var url = 'https://desolate-anchorage-29312.herokuapp.com/generateTxnToken';

        var body = json.encode({
          "mid": mid,
          "key_secret": PAYTM_MERCHANT_KEY,
          "website": website,
          "orderId": orderId,
          "amount": amount.toString(),
          "callbackUrl": callBackUrl,
          "custId": "122",
          "mode": mode.toString(),
          "testing": testing ? 0 : 1
        });

        try {
          final response = await http.post(
            url,
            body: body,
            headers: {'Content-type': "application/json"},
          );
          print("Response is");
          print(response.body);
          String txnToken = response.body;
          setState(() {
            payment_response = txnToken;
          });

          var paytmResponse = Paytm.payWithPaytm(
              mid, orderId, txnToken, amount.toString(), callBackUrl, testing);

          paytmResponse.then((value) {
            print(value);
            setState(() {
              loading = false;
              print("Value is ");
              print(value);
              if(value['error']){
                payment_response = value['errorMessage'];
              }else{
                if(value['response']!=null){
                  payment_response = value['response']['STATUS'];
                }
              }
              payment_response += "\n"+value.toString();
            });
          });
        } catch (e) {
          print(e);
        }
  }
  
  ``` 

## GENERATE TOKEN
For SERVER CODE:
[Paytm Plugin Server Code](https://github.com/mrdishant/Paytm-Plugin-Server)

## Support
<a href="https://www.buymeacoffee.com/mrdishant" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" style="height: 30px !important;width: 117px !important;" ></a>

For Cloning the example app code visit:
[Paytm Plugin](https://github.com/mrdishant/Paytm-All-in-one-flutter-sdk.git)

For any query :
Mail me at mr.dishantmahajan@gmail.com

CHANGELOG.md
## 3.0.1

* Update: Paytm iOS and Android SDK updated.


## 3.0.0

* Update: Paytm iOS SDK updated + Null Safety Added.


## 2.0.0

* Update: Paytm iOS SDK updated + Stagging Support Added.
* Bug Fix: Paytm SDK Response Bug Fix.


## 1.0.8

* Update: Paytm Android SDK updated.
* Bug Fix: Paytm Android SDK Response Fix.

## 1.0.7

* Update: Paytm Android SDK updated.

## 1.0.6

* Update: Updated Min SDK Version Android.

## 1.0.5

* Update: Updated ReadMe file.

## 1.0.4

* Update: Added Staging Support (Android).

## 1.0.3

* Update: Bug Fixes.

## 1.0.2

* Update: PAYTM ALL IN ONE SDK.

## 0.0.5

* Update: Bug Fixes.

## 0.0.4

* Update: ANDROID X Support added.

## 0.0.3

* Update: ANDROID X Support added.

## 0.0.2

* Update: Bug Fixes.

## 0.0.1

* Update: Paytm as a payment gateway in your flutter app.

example/lib/main.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:paytm/paytm.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String payment_response;

  //Live
  String mid = "LIVE_MID_HERE";
  String PAYTM_MERCHANT_KEY = "LIVE_KEY_HERE";
  String website = "DEFAULT";
  bool testing = false;

  //Testing
  // String mid = "TEST_MID_HERE";
  // String PAYTM_MERCHANT_KEY = "TEST_KEY_HERE";
  // String website = "WEBSTAGING";
  // bool testing = true;

  double amount = 1;
  bool loading = false;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Paytm example app'),
        ),
        body: SingleChildScrollView(
          child: Padding(
            padding: EdgeInsets.all(10.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Text(
                    'For Testing Credentials make sure appInvokeEnabled set to FALSE or Paytm APP is not installed on the device.'),

                SizedBox(
                  height: 10,
                ),

                TextField(
                  onChanged: (value) {
                    mid = value;
                  },
                  decoration: InputDecoration(hintText: "Enter MID here"),
                  keyboardType: TextInputType.text,
                ),
                TextField(
                  onChanged: (value) {
                    PAYTM_MERCHANT_KEY = value;
                  },
                  decoration:
                      InputDecoration(hintText: "Enter Merchant Key here"),
                  keyboardType: TextInputType.text,
                ),
                TextField(
                  onChanged: (value) {
                    website = value;
                  },
                  decoration: InputDecoration(
                      hintText: "Enter Website here (Probably DEFAULT)"),
                  keyboardType: TextInputType.text,
                ),
                TextField(
                  onChanged: (value) {
                    try {
                      amount = double.tryParse(value);
                    } catch (e) {
                      print(e);
                    }
                  },
                  decoration: InputDecoration(hintText: "Enter Amount here"),
                  keyboardType: TextInputType.number,
                ),
                SizedBox(
                  height: 10,
                ),
                payment_response != null
                    ? Text('Response: $payment_response\n')
                    : Container(),
//                loading
//                    ? Center(
//                        child: Container(
//                            width: 50,
//                            height: 50,
//                            child: CircularProgressIndicator()),
//                      )
//                    : Container(),
                RaisedButton(
                  onPressed: () {
                    //Firstly Generate CheckSum bcoz Paytm Require this
                    generateTxnToken(0);
                  },
                  color: Colors.blue,
                  child: Text(
                    "Pay using Wallet",
                    style: TextStyle(color: Colors.white),
                  ),
                ),
                RaisedButton(
                  onPressed: () {
                    //Firstly Generate CheckSum bcoz Paytm Require this
                    generateTxnToken(1);
                  },
                  color: Colors.blue,
                  child: Text(
                    "Pay using Net Banking",
                    style: TextStyle(color: Colors.white),
                  ),
                ),
                RaisedButton(
                  onPressed: () {
                    //Firstly Generate CheckSum bcoz Paytm Require this
                    generateTxnToken(2);
                  },
                  color: Colors.blue,
                  child: Text(
                    "Pay using UPI",
                    style: TextStyle(color: Colors.white),
                  ),
                ),
                RaisedButton(
                  onPressed: () {
                    //Firstly Generate CheckSum bcoz Paytm Require this
                    generateTxnToken(3);
                  },
                  color: Colors.blue,
                  child: Text(
                    "Pay using Credit Card",
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  void generateTxnToken(int mode) async {
    setState(() {
      loading = true;
    });
    String orderId = DateTime.now().millisecondsSinceEpoch.toString();

    String callBackUrl = (testing
            ? 'https://securegw-stage.paytm.in'
            : 'https://securegw.paytm.in') +
        '/theia/paytmCallback?ORDER_ID=' +
        orderId;

    //Host the Server Side Code on your Server and use your URL here. The following URL may or may not work. Because hosted on free server.
    //Server Side code url: https://github.com/mrdishant/Paytm-Plugin-Server

    var url = 'https://desolate-anchorage-29312.herokuapp.com/generateTxnToken';

    var body = json.encode({
      "mid": mid,
      "key_secret": PAYTM_MERCHANT_KEY,
      "website": website,
      "orderId": orderId,
      "amount": amount.toString(),
      "callbackUrl": callBackUrl,
      "custId": "122",
      "mode": mode.toString(),
      "testing": testing ? 0 : 1
    });

    try {
      final response = await http.post(
        Uri.parse(url),
        body: body,
        headers: {'Content-type': "application/json"},
      );

      print("Response is");
      print(response.body);
      String txnToken = response.body;
      setState(() {
        payment_response = txnToken;
      });

      var paytmResponse = Paytm.payWithPaytm(
          mId: mid,
          orderId: orderId,
          txnToken: txnToken,
          txnAmount: amount.toString(),
          callBackUrl: callBackUrl,
          staging: testing,
          appInvokeEnabled: false);

      paytmResponse.then((value) {
        print(value);
        setState(() {
          loading = false;
          print("Value is ");
          print(value);
          if (value['error']) {
            payment_response = value['errorMessage'];
          } else {
            if (value['response'] != null) {
              payment_response = value['response']['STATUS'];
            }
          }
          payment_response += "\n" + value.toString();
        });
      });
    } catch (e) {
      print(e);
    }
  }
}
